### ISSUE TITLE
AutoEncoder in PyTorch/Python

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#655]
Some of my Java code makes use of Python code for deep learning. Since we are having trouble with the DL4J autoencoder #635 , it might be worth considering a Python/Pytorch one instead. See this link:
https://medium.com/pytorch/implementing-an-autoencoder-in-pytorch-19baa22647d1

This walks through setting up a Python autoencoder for MNIST. If we can get this to work, we might be able to train one on our Picbreeder images.

I would recommend working on this Python code in a completely separate directory that is not part of the MM-NEAT code at first. Once you get something working, then there is a place in the MM-NEAT code where we can move/add the Python code.

By the way, there are many other Pytorch autoencoder examples out there, so feel free to look around. Here is another one:
https://github.com/L1aoXingyu/pytorch-beginner/tree/master/08-AutoEncoder

**Comment by annazwicker** [comment]
Kept the autoencoder from this link: https://github.com/L1aoXingyu/pytorch-beginner/tree/master/08-AutoEncoder.  Seems to run, but very slowly (uses MNIST).  Other link was useful for information but code wouldn't run as easily and wasn't kept for future use.

**Comment by schrum2** [comment]
If this ends up working, then the next step is to try making it work with some picbreeder image data.

**Comment by annazwicker** [comment]
This link seems to be specific to what we are looking to do, but we might have to reformat the folder with the training images.  I haven't done enough in Python to know how this is really working, but I think we would need to specify additional information for each image.
https://towardsdatascience.com/how-to-load-a-custom-image-dataset-on-pytorch-bf10b2c529e0
I checked several other links, but this one seems to be the most specific to what we are trying to do with the PictureTargetTask images.

**Comment by schrum2** [comment]
It looks like you may have to run `pip install --user pandas` in the Anaconda prompt for this to work. Ask @MBatt1 if you run into problems with this. Here are some other links that may be helpful.

http://www.interstellarengine.com/pytorch/DATA-LOADING-BATCHING-and-SHUFFLING-Auto-Encoder-with-dictionary-sample-in-output-TUTORIAL.html
https://stackoverflow.com/questions/45099554/how-to-simplify-dataloader-for-autoencoder-in-pytorch
https://github.com/utkuozbulak/pytorch-custom-dataset-examples

**Comment by annazwicker** [comment]
The PyTorch Autoencoder produces images that somewhat resemble the skull image (closer than any other attempt).  The images produced appear to be faded (almost like they have an opaque screen over them, but the original shape is still visible), though we are not sure why.  Examples can be seen below:

![image_450](https://user-images.githubusercontent.com/82399865/120393557-bd165480-c2f7-11eb-9ae5-5ebc2dc0d2f7.png)
![image_480](https://user-images.githubusercontent.com/82399865/120393559-bdaeeb00-c2f7-11eb-901c-346f605c3af4.png)
![image_490](https://user-images.githubusercontent.com/82399865/120393560-bdaeeb00-c2f7-11eb-8a83-605e9cbba04a.png)

**Comment by schrum2** [comment]
Find a way to save the autoencoder trained by this code. Look at the end of `src\main\python\GAN\main.py` for the `torch.save` command. This should create a `pth` file containing the network.

**Comment by schrum2** [comment]
Also, once you can save the pth file as described above, you need to make a separate Python script capable of loading the trained network weights from the pth file so that it can be used to process new inputs. Specifically, this script will launch, load the network, and then wait for inputs from the console in the format of a long flat list of floating point values. The script will then parse that list into a PyTorch tensor, send it as input to the autoencoder, and print the output of the autoencoder back to the console as a list of floating point numbers. 

This is similar to code in `src\main\python\GAN\generator_ws.py`

Note that you may need to more the `autoencoder` class out of `MyAutoencoder.py` and put it in a common class so that it can be imported both into that script and the new one you write.

**Comment by schrum2** [comment]
The PyTorch autoencoder works, so I guess we'll close this issue. Additional work related to this will be documented in issue #642
